app/models/manageiq/providers/ibm_cloud/inventory/persister/vpc.rb

```diff
@@ -42,4 +45,11 @@ def initialize_storage_inventory_collections
     add_storage_collection(:cloud_volumes)
     add_storage_collection(:cloud_volume_types)
   end
+
+  # Automatically add the local model class and id of ems to the database record.
+  # @return [void]
+  def add_resource_group(builder)
+    builder.add_properties(:model_class => ::ManageIQ::Providers::IbmCloud::VPC::ResourceGroup)
+    builder.add_default_values(:ems_id => manager.id)
```

Comment:
So this will be the third redefinition of the resource_group inventory collection by a provider.  I think it is time to move this into the core persister definition: https://github.com/ManageIQ/manageiq/pull/21196

Once that is merged this just becomes `add_cloud_collection(:resource_groups)` like the rest of the collections.

Reply:
Merged

Reply:
When I try to integrate your change I find:
1.  Removing the `resource_groups` relation in the CloudManager results in a NoMethod exception during refresh.
2. With the requested change the persister finishes, but a blank type is inserted into the DB.
3. If I move the ResourceGroup class into the CloudManager namespace, with the requisite file move, then everything gets persisted properly.

Can you fix up the core CloudManager so that I can remove those components? Please. The piece you changed wasn't what made it difficult to persist the resource groups. It was the missing relation in CloudManager and the unhelpful exception message. Also the persister shouldn't be saving nil for type. It should either be an exception with instructions on how to fix, or, if available, a default could be used. I doubt that there is any functional difference in an empty ResourceGroup class definition under the plugin CloudManager and using the parent.

Reply:
> Removing the resource_groups relation in the CloudManager results in a NoMethod exception during refresh.

I only changed the core persister, you still need this for now

> With the requested change the persister finishes, but a blank type is inserted into the DB.

Did you update core?  It sets the model_class for you.

> If I move the ResourceGroup class into the CloudManager namespace, with the requisite file move, then everything gets persisted properly.

That's up to you, this is what I need to do for the other two providers but they need a data migration to fixup existing records.  If you go that route you'll have to set the model_class manually.

This is what I'm suggesting: https://github.com/agrare/manageiq-providers-ibm_cloud/commit/83eb56a6b387a06348a97d0e099a0209828a6e01

It explicitly checks the type that gets created so it isn't `nil`, https://travis-ci.com/github/agrare/manageiq-providers-ibm_cloud/jobs/501670758

Reply:
I'm saying your core change is a partial solution, and hides some of the steps to get it to work properly. Having `add_cloud_collection(:resource_groups)` gives the impression that it is a full solution when it isn't. This PR is a full solution and also is followable by someone looking to implement resource_groups. Having the hidden gotchas as listed in my previous comment is not optimal.

I have the spec updated with a 'type' check but I like your implementation better.

Reply:
> I'm saying your core change is a partial solution, and hides some of the steps to get it to work properly

Sorry I'm not sure what you mean, you mean it gives the impression that you don't need the association?  That wasn't what I was suggesting, my bad if that's how it sounded.

That said, idk who wrote this resource group stuff originally but there are a ton of issues with it (the base ResourceGroup doesn't even have a `belongs_to :ext_management_system` even though it has an `:ems_id`).  I'm tackling the proper namespacing in a different patch set due to the schema change required, I'll add the associations to core as part of that.

Reply:
> Did you update core? It sets the model_class for you.
Yes, I did and removed the pieces I thought should be covered. I then readded the pieces that were missing and noted my observations. Like without a defined `ManageIQ::Providers::IbmCloud::VPC::CloudManager::ResourceGroup` the type is saved as nil.

Reply:
> Sorry I'm not sure what you mean, you mean it gives the impression that you don't need the association? That wasn't what I was suggesting, my bad if that's how it sounded.

Yes, that is what I was saying. The other associations don't need explicit relations added to the CloudManager to work. So when I see `add_cloud_collection(:resource_groups)` I assume that core is taking care of everything. This is currently not the case.